src/main/java/com/wavefront/sdk/grpc/WavefrontServerTracerFactory.java

```diff
@@ -0,0 +1,299 @@
+        put(CLUSTER_TAG_KEY, applicationTags.getCluster() == null ? NULL_TAG_VAL :
+            applicationTags.getCluster());
+        put(SERVICE_TAG_KEY, applicationTags.getService());
+        put(SHARD_TAG_KEY, applicationTags.getShard() == null ? NULL_TAG_VAL :
+            applicationTags.getShard());
+        put(GRPC_SERVICE_TAG_KEY, grpcService);
+      }};
+      this.overallAggregatedPerSourceTags = ImmutableMap.<String, String>builder().
+          put(CLUSTER_TAG_KEY, applicationTags.getCluster() == null ? NULL_TAG_VAL :
+              applicationTags.getCluster()).
+          put(SERVICE_TAG_KEY, applicationTags.getService()).
+          put(SHARD_TAG_KEY, applicationTags.getShard() == null ? NULL_TAG_VAL :
+              applicationTags.getShard()).
+          build();
+      // update requests inflight
+      getGaugeValue(new MetricName(REQUEST_PREFIX + methodName + ".inflight", allTags)).
+          incrementAndGet();
+      getGaugeValue(new MetricName("server.total_requests.inflight",
+          overallAggregatedPerSourceTags)).incrementAndGet();
+    }
+
+    @Override
+    public void serverCallStarted(ServerCallInfo<?, ?> callInfo) {
+      streamingMethod.set(Utils.isStreamingMethod(callInfo.getMethodDescriptor().getType()));
+    }
+
+    @Override
+    public void outboundWireSize(long bytes) {
+      responseBytes.addAndGet(bytes);
+    }
+
+    @Override
+    public void inboundWireSize(long bytes) {
+      requestBytes.addAndGet(bytes);
+    }
+
+    @Override
+    public void inboundMessageRead(int seqNo, long optionalWireSize,
+                                   long optionalUncompressedSize) {
+      if (shouldRecordStreamingStats()) {
+        requestMessageCount.incrementAndGet();
+        wfGrpcReporter.incrementCounter(
+            new MetricName(REQUEST_PREFIX + methodName + ".streaming.messages.count", allTags));
+        if (optionalWireSize >= 0) {
+          wfGrpcReporter.updateHistogram(
+              new MetricName(REQUEST_PREFIX + methodName + ".streaming.message_bytes", allTags),
+              optionalWireSize);
+        }
+      }
+    }
+
+    @Override
+    public void outboundMessageSent(int seqNo, long optionalWireSize,
+                                    long optionalUncompressedSize) {
+      if (shouldRecordStreamingStats()) {
+        responseMessageCount.incrementAndGet();
+        wfGrpcReporter.incrementCounter(
+            new MetricName(RESPONSE_PREFIX + methodName + ".streaming.messages.count", allTags));
+        if (optionalWireSize >= 0) {
+          wfGrpcReporter.updateHistogram(
+              new MetricName(RESPONSE_PREFIX + methodName + ".streaming.message_bytes", allTags),
+              optionalWireSize);
+        }
+      }
+    }
+
+    @Override
+    public void streamClosed(Status status) {
+      if (streamClosed.getAndSet(true)) {
+        return;
+      }
+      long rpcLatency = System.currentTimeMillis() - startTime;
```

Comment:
If this method it is essentially the same for the client and server (apart from the metric prefix), Could this be refactored into a common method and reused?

Reply:
Not sure if we can do this in a simple way as its using lot of other member variables (startTime, req/resp bytes, etc.).